### ISSUE TITLE
Improve error messages for better UX

### ISSUE THREAD

**Issue by jvican** [scalacenter/bloop#533]
Bloop has decent error messages, but I'd like to strive for excellent error messages. An error message is excellent if describes the issue at hand shortly and gives actionable feedback to the user.
  
The idea of this ticket is that some present or future Bloop contributor goes through all the instances of `state.logger.error` or `state.logger.info` or `state.logger.warn` and derivatives to find which error messages could be improved.
  
In general, error messages that start with "Expected XXX, obtained YYY" or "Missing bla bla bla" are the best. Ideally, we'd like all messages to stay consistent with each other using the same set of recipes.
  
All the nice error messages could be created in an object `Feedback` that is accessible from all places in `frontend`. New ideas are welcome!

**Comment by tindzk** [comment]
We could also use colours and format the messages if these features are supported by the terminal. For example:

> Compiling 'project'...

With italics, the message would be more concise and readable:

> Compiling _project_...

Recent terminals even support RGB colours, so we can even use a custom colour scheme for [better readability](https://www.viget.com/articles/color-contrast/). [Here](https://asciinema.org/a/0urh4T5dDimpFF6gA7EbdjiSw?t=8) is an example for inspiration ([code](https://github.com/sparsetech/instructor/blob/master/src/main/scala/instructor/Ansi.scala)).

**Comment by jvican** [comment]
Yes, I also had this in mind. We'll soon have the project name in the compilation logs, so we'll be able to implement it. In general, I think it would be useful to add all the infrastructure to change messages gradually.
  
We should probably use `fansi` or something similar that is supported in Windows. Or even support Windows in fansi and just depend on that. I would prefer to use a widely used community library and that we also have an impact upstream improving things as we need them.

**Comment by tindzk** [comment]
fansi looks good. We could also show progress bars for each project like in [cargo-fancy](https://github.com/alexcrichton/cargo-fancy).

**Comment by jvican** [comment]
Man, that's pretty cool. I would be super happy to have such thing in our UI (we can support it since we have access to the compiler progress notifications).
  
Do you know what happens if there's a compiler error? The asciicast in https://asciinema.org/a/dbid3w8jqdpc7sauqdzkn784t doesn't show how they are displayed to the user.

**Comment by tindzk** [comment]
I did not try it, but I presume it would print the errors above the progress bars. Does Bloop/Zinc compile project dependencies in parallel?

**Comment by jvican** [comment]
Yeah it does. Thanks for the links, let me know if you wanna tackle this issue or r otherwise I’ll see if I can have a look at it next week.

**Comment by tindzk** [comment]
Feel free to look into it. I will focus on consolidating Scala.js support for now and in a few weeks could help with the UX improvements.

**Comment by tindzk** [comment]
We should also improve the UX of the CLI parameter parsing. For example, when I pass in a wrong value to the optimisation parameter, it should give a suggestion or list all valid options. The current message is not particularly helpful:

> Unrecognized optimizer config: production

Recently, I have started using [argyle](https://github.com/jfkelley/argyle) which is more flexible than caseapp and does not suffer from the issues I reported here before.
